Json format bails too early

I might have found an issue with how the `--bail` and `--json` flags interact with each other.

I have created a minimal repo that consistently reproduces the issue: https://github.com/dreamorosi/validate-repro

Given the following `package.json`:

```json
{
  "name": "test-package",
  "version": "1.0.0",
  "exports": {
    ".": {
      "require": {
        "types": "./lib/cjs/foo.d.ts",
        "default": "./lib/cjs/foo.js"
      },
      "import": {
        "types": "./lib/esm/foo.d.ts",
        "default": "./lib/esm/foo.mjs"
      }
    },
    "./non-existent": {
      "import": "./lib/esm/bar.mjs",
      "require": "./lib/cjs/bar.js"
    }
  },
  "dependencies": {
    "validate-package-exports": "^0.2.0"
  }
}
```

with the following file structure:

```bash
lib/
├─ esm/
│  ├─ foo.mjs/
│  ├─ foo.d.ts
├─ cjs/
│  ├─ foo.js
│  ├─ foo.d.ts
```

The package should report at least two errors since the exports under `./non-existent` are incorrect.

When using the `--bail` option alone, the package correctly reports two errors and stops.

```bash
npx validate-package-exports --check --verify --bail       
❌ file-exists: lib/esm/bar.mjs does not exist (["exports","./non-existent","import"]) Error: ENOENT: no such file or directory, stat '/Users/andre/Sites/validate-repro/lib/esm/bar.mjs'
❌ file-exists: lib/cjs/bar.js does not exist (["exports","./non-existent","require"]) Error: ENOENT: no such file or directory, stat '/Users/andre/Sites/validate-repro/lib/cjs/bar.js'
```


However when doing the same but with the `--json` option, the package reports only the first error and stops.

```bash
npx validate-package-exports --check --verify --bail --json
[
  {
    "name": "file-exists",
    "code": 1,
    "message": "lib/esm/bar.mjs does not exist",
    "entryPoint": {
      "moduleName": "test-package/non-existent",
      "packagePath": "/Users/andre/Sites/validate-repro/package.json",
      "type": "module",
      "fileName": "bar.mjs",
      "relativePath": "lib/esm/bar.mjs",
      "directory": "/Users/andre/Sites/validate-repro/lib/esm",
      "resolvedPath": "/Users/andre/Sites/validate-repro/lib/esm/bar.mjs",
      "subpath": "./non-existent",
      "condition": "import",
      "itemPath": [
        "exports",
        "./non-existent",
        "import"
      ]
    },
    "error": {
      "errno": -2,
      "code": "ENOENT",
      "syscall": "stat",
      "path": "/Users/andre/Sites/validate-repro/lib/esm/bar.mjs"
    }
  }
]        
```

On second thought I don't know which one of the two behaviors is the correct one. I am not sure what you intended for the `--bail` flag and if you wanted it to stop after the first error, or stop on the first error of a specific export.

From my side I think the latter is better than the first, however one could also make the case for changing some of the behavior of the validation altogether. 

For example, if the file is not found (like in this case), then it doesn't make sense to do other types of validations (i.e. imports) since we know they'll always fail. For example, in this case, without the `--bail` flag there would be many errors that are redundant (only the first two are relevant imo):

```bash
❌ file-exists: lib/esm/bar.mjs does not exist (["exports","./non-existent","import"]) Error: ENOENT: no such file or directory, stat '/Users/andre/Sites/validate-repro/lib/esm/bar.mjs'
❌ file-exists: lib/cjs/bar.js does not exist (["exports","./non-existent","require"]) Error: ENOENT: no such file or directory, stat '/Users/andre/Sites/validate-repro/lib/cjs/bar.js'
❌ check-syntax: Could not validate syntax for lib/esm/bar.mjs (["exports","./non-existent","import"]) Error: Command failed: node --check /Users/andre/Sites/validate-repro/lib/esm/bar.mjs
node:internal/modules/cjs/loader:1147
  throw err;
  ^

Error: Cannot find module '/Users/andre/Sites/validate-repro/lib/esm/bar.mjs'
    at Module._resolveFilename (node:internal/modules/cjs/loader:1144:15)
    at node:internal/main/check_syntax:33:20 {
  code: 'MODULE_NOT_FOUND',
  requireStack: []
}

Node.js v20.11.1
❌ check-syntax: Could not validate syntax for lib/cjs/bar.js (["exports","./non-existent","require"]) Error: Command failed: node --check /Users/andre/Sites/validate-repro/lib/cjs/bar.js
node:internal/modules/cjs/loader:1147
  throw err;
  ^

Error: Cannot find module '/Users/andre/Sites/validate-repro/lib/cjs/bar.js'
    at Module._resolveFilename (node:internal/modules/cjs/loader:1144:15)
    at node:internal/main/check_syntax:33:20 {
  code: 'MODULE_NOT_FOUND',
  requireStack: []
}

Node.js v20.11.1
❌ require: test-package/non-existent cannot be required (["exports","./non-existent","require"]) Error: Command failed: node --input-type=commonjs --eval="require('test-package/non-existent');"
node:internal/modules/cjs/loader:565
    throw e;
    ^

Error: Cannot find module '/Users/andre/Sites/validate-repro/lib/cjs/bar.js'
    at createEsmNotFoundErr (node:internal/modules/cjs/loader:1181:15)
    at finalizeEsmResolution (node:internal/modules/cjs/loader:1169:15)
    at trySelf (node:internal/modules/cjs/loader:558:12)
    at Module._resolveFilename (node:internal/modules/cjs/loader:1121:24)
    at Module._load (node:internal/modules/cjs/loader:985:27)
    at Module.require (node:internal/modules/cjs/loader:1235:19)
    at require (node:internal/modules/helpers:176:18)
    at [eval]:1:1
    at runScriptInThisContext (node:internal/vm:144:10)
    at node:internal/process/execution:109:14 {
  code: 'MODULE_NOT_FOUND',
  path: '/Users/andre/Sites/validate-repro/package.json'
}

Node.js v20.11.1
❌ import: test-package/non-existent cannot be imported (["exports","./non-existent","import"]) Error: Command failed: node --input-type=module --eval="import 'test-package/non-existent';"
node:internal/modules/esm/resolve:264
    throw new ERR_MODULE_NOT_FOUND(
          ^

Error [ERR_MODULE_NOT_FOUND]: Cannot find module '/Users/andre/Sites/validate-repro/lib/esm/bar.mjs' imported from /Users/andre/Sites/validate-repro/[eval1]
    at finalizeResolution (node:internal/modules/esm/resolve:264:11)
    at moduleResolve (node:internal/modules/esm/resolve:917:10)
    at defaultResolve (node:internal/modules/esm/resolve:1130:11)
    at ModuleLoader.defaultResolve (node:internal/modules/esm/loader:396:12)
    at ModuleLoader.resolve (node:internal/modules/esm/loader:365:25)
    at ModuleLoader.getModuleJob (node:internal/modules/esm/loader:240:38)
    at ModuleWrap.<anonymous> (node:internal/modules/esm/module_job:85:39)
    at link (node:internal/modules/esm/module_job:84:36) {
  code: 'ERR_MODULE_NOT_FOUND',
  url: 'file:///Users/andre/Sites/validate-repro/lib/esm/bar.mjs'
}

Node.js v20.11.1
```

--- a/src/lib/Validator.ts
+++ b/src/lib/Validator.ts
@@ -48,51 +48,71 @@ export class Validator extends EventEmitter {
     [results].flat().forEach(result => this.processResult(result));
   }
 
-  protected async checkFilesExist(entryPoints: EntryPoint[]): Promise<void> {
-    await Readable.from(entryPoints).forEach(
-      async (entryPoint: EntryPoint) => {
-        const results = await checkFileExists(entryPoint);
+  protected async checkFilesExist(entryPoints: EntryPoint[]): Promise<Result[]> {
+    return await Readable.from(entryPoints)
+      .map(
+        async (entryPoint: EntryPoint) => {
+          const results = await checkFileExists(entryPoint);
 
-        this.processResults(results);
-      },
-      {
+          this.processResults(results);
+
+          return results;
+        },
+        {
+          signal: this.#controller.signal,
+          concurrency: this.options.concurrency,
+        },
+      )
+      .toArray({
         signal: this.#controller.signal,
-        concurrency: this.options.concurrency,
-      },
-    );
+      });
   }
 
-  protected async checkSyntax(entryPoints: EntryPoint[]): Promise<void> {
+  protected async checkSyntax(entryPoints: EntryPoint[]): Promise<Result[]> {
     const jsEntryPoints = entryPoints.filter(entryPoint => /\.[cm]?js$/i.test(entryPoint.resolvedPath));
 
-    await Readable.from(jsEntryPoints).forEach(
-      async (entryPoint: EntryPoint) => {
-        const results = await checkSyntax(entryPoint, { signal: this.#controller.signal });
+    return await Readable.from(jsEntryPoints)
+      .map(
+        async (entryPoint: EntryPoint) => {
+          const results = await checkSyntax(entryPoint, { signal: this.#controller.signal });
 
-        this.processResults(results);
-      },
-      {
+          this.processResults(results);
+
+          return results;
+        },
+        {
+          signal: this.#controller.signal,
+          concurrency: this.options.concurrency,
+        },
+      )
+      .toArray({
         signal: this.#controller.signal,
-        concurrency: this.options.concurrency,
-      },
-    );
+      });
   }
 
-  protected async verifyIncludes(entryPoints: EntryPoint[]): Promise<void> {
-    await Readable.from(entryPoints).forEach(
-      async (entryPoint: EntryPoint) => {
-        const results = await verifyEntryPoint(entryPoint, {
-          cwd: this.packageDirectory,
-          signal: this.#controller.signal,
-        });
+  protected async verifyIncludes(entryPoints: EntryPoint[]): Promise<Result[]> {
+    return (
+      await Readable.from(entryPoints)
+        .map(
+          async (entryPoint: EntryPoint) => {
+            const results = await verifyEntryPoint(entryPoint, {
+              cwd: this.packageDirectory,
+              signal: this.#controller.signal,
+            });
 
-        this.processResults(results);
-      },
-      {
-        signal: this.#controller.signal,
-        concurrency: this.options.concurrency,
-      },
-    );
+            this.processResults(results);
+
+            return results;
+          },
+          {
+            signal: this.#controller.signal,
+            concurrency: this.options.concurrency,
+          },
+        )
+        .toArray({
+          signal: this.#controller.signal,
+        })
+    ).flat();
   }
 
   async run(): Promise<ExitCode> {
@@ -111,18 +131,34 @@ export class Validator extends EventEmitter {
       signal: this.#controller.signal,
     });
 
+    const entryPointsWithErrors = new Set<EntryPoint>();
+
+    const processResults = (results: Result[]): void => {
+      results.forEach(result => {
+        if (result.code === ResultCode.Error) {
+          entryPointsWithErrors.add(result.entryPoint);
+        }
+      });
+    };
+
+    const getNextEntryPoints = (entryPoints: EntryPoint[]): EntryPoint[] => {
+      return entryPoints.filter(entryPoint => !entryPointsWithErrors.has(entryPoint));
+    };
+
     // Always check to see if the file exists.
-    await this.checkFilesExist(entryPoints);
+    processResults(await this.checkFilesExist(entryPoints));
 
     // Optionally run a syntax check.
     if (this.options.check) {
-      await this.checkSyntax(entryPoints);
+      processResults(await this.checkSyntax(getNextEntryPoints(entryPoints)));
     }
 
     // Optionally try to `import`/`require` the module.
     if (this.options.verify) {
-      await this.verifyIncludes(entryPoints);
+      processResults(await this.verifyIncludes(getNextEntryPoints(entryPoints)));
     }
+
+    // TODO: maybe do some reporting using `entryPointsWithErrors`
 
     return this.#exitCode;
   }

